rust/windows-client/src-tauri/src/gui.rs

```diff
@@ -0,0 +1,518 @@
+                tracing::debug!("got {} resources", resources.len());
+            }
+        }
+    }
+    tracing::debug!("GUI controller task exiting cleanly");
+    Ok(())
+}
+
+pub(crate) struct AuthCallback {
+    token: SecretString,
+    _identifier: SecretString,
+}
+
+fn parse_auth_callback(input: &SecretString) -> Result<AuthCallback> {
+    use secrecy::ExposeSecret;
+
+    let url = url::Url::parse(input.expose_secret())?;
+
+    let mut token = None;
+    let mut identifier = None;
+
+    for (key, value) in url.query_pairs() {
+        match key.as_ref() {
+            "client_auth_token" => {
+                if token.is_some() {
+                    bail!("client_auth_token must appear exactly once");
+                }
+                token = Some(SecretString::new(value.to_string()));
+            }
+            "identity_provider_identifier" => {
+                if identifier.is_some() {
+                    bail!("identity_provider_identifier must appear exactly once");
+                }
+                identifier = Some(SecretString::new(value.to_string()));
+            }
+            _ => {}
+        }
+    }
+
+    Ok(AuthCallback {
+        token: token.ok_or_else(|| anyhow!("expected client_auth_token"))?,
+        _identifier: identifier.ok_or_else(|| anyhow!("expected identity_provider_identifier"))?,
+    })
+}
+
+/// The information needed for the GUI to display a resource inside the Firezone VPN
+struct _ResourceDisplay {
+    // TODO: The ID in ResourceDescription is something like a number that's safe to use internally, right? Like it won't have slashes or emojis or any odd characters that a name or URL might have?
+    id: String,
+    /// User-friendly name, e.g. "GitLab"
+    name: String,
+    /// What will be copied to the clipboard to paste into a web browser
+    url: Url,
+}
+
+fn _signed_in_menu(user_email: &str, resources: &[_ResourceDisplay]) -> SystemTrayMenu {
+    let mut menu = SystemTrayMenu::new()
+        .add_item(
+            CustomMenuItem::new("".to_string(), format!("Signed in as {user_email}")).disabled(),
+        )
+        .add_item(CustomMenuItem::new("/sign_out".to_string(), "Sign out"))
+        .add_native_item(SystemTrayMenuItem::Separator)
+        .add_item(CustomMenuItem::new("".to_string(), "Resources").disabled());
+
+    for _ResourceDisplay { id, name, url } in resources {
+        let submenu = SystemTrayMenu::new().add_item(CustomMenuItem::new(
+            format!("/resource/{id}"),
+            url.to_string(),
+        ));
+        menu = menu.add_submenu(SystemTraySubmenu::new(name, submenu));
+    }
+
+    menu = menu
+        .add_native_item(SystemTrayMenuItem::Separator)
+        .add_item(CustomMenuItem::new("/about".to_string(), "About"))
+        .add_item(CustomMenuItem::new("/settings".to_string(), "Settings"))
+        .add_item(CustomMenuItem::new("/quit".to_string(), "Quit Firezone").accelerator("Ctrl+Q"));
```

Comment:
```suggestion
        .add_item(CustomMenuItem::new("/quit".to_string(), "Disconnect and quit Firezone").accelerator("Ctrl+Q"));
```

Reply:
I accepted the text change and #2812 is open to fix the actual behavior